features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c

```diff
@@ -0,0 +1,256 @@
+          memset ( temp_buf, 0 , sizeof(temp_buf) );
+
+          ret = convert_CrysError_to_mbedtls_err( CRYS_COMMON_ConvertLswMswWordsToMsbLsbBytes( temp_buf, (grp->nbits+7)/8,
+                                                            kgParams->privKey.PrivKeyDbBuff,
+                                                            4*((((grp->nbits+7)/8)+3)/4) ) );
+          if( ret != 0 )
+          {
+              mbedtls_zeroize( temp_buf, sizeof( temp_buf ) );
+              goto cleanup;
+          }
+
+          MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary( d, temp_buf, (grp->nbits+7)/8 ) );
+          mbedtls_zeroize( temp_buf, sizeof( temp_buf ) );
+      }
+
+      /* if CRYS_ECPKI_GetEcDomain returns NULL, than the given curve is wither Montgomery 25519
+       * or another curve which is not supported by CC310*/
+      else if ( grp->id ==  MBEDTLS_ECP_DP_CURVE25519 )
+      {
+          size_t priv_key_size =  public_key_size = CURVE_25519_KEY_SIZE ;
+
+          cc_ecc_25519_keygen_params_t* kgParams =  mbedtls_calloc( 1, sizeof(cc_ecc_25519_keygen_params_t) );
+
+          if ( kgParams == NULL )
+              return ( MBEDTLS_ERR_ECP_ALLOC_FAILED );
+          pHeap = ( uint8_t* )kgParams;
+          heapSize = sizeof(cc_ecc_25519_keygen_params_t);
+
+          ret = convert_CrysError_to_mbedtls_err( CRYS_ECMONT_KeyPair( kgParams->pubKey, ( size_t* )&public_key_size, kgParams->privKey,
+                                    &priv_key_size, &cc_rand, convert_mbedtls_to_cc_rand,
+                                    &kgParams->kgTempData ) );
+          if( ret != 0 )
+          {
+              goto cleanup;
+          }
+
+          MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( d, kgParams->privKey, priv_key_size ) );
+          MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary(  &Q->X, kgParams->pubKey, public_key_size ) );
+          MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &Q->Z, 1 ) );
+      }
+      else
+          ret =  MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+
+cleanup:
+
+    if ( pHeap )
+    {
+        mbedtls_zeroize( pHeap, heapSize );
+        mbedtls_free( pHeap );
+    }
+
+    return ( ret );
+}
+#endif /* MBEDTLS_ECDH_GEN_PUBLIC_ALT */
+
+/*
+ * Compute shared secret (SEC1 3.3.1)
+ */
+#if defined (MBEDTLS_ECDH_COMPUTE_SHARED_ALT)
+int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
+                         const mbedtls_ecp_point *Q, const mbedtls_mpi *d,
+                         int (*f_rng)(void *, unsigned char *, size_t),
+                         void *p_rng )
+{
+    int ret;
+    void* pHeap = NULL;
+    size_t   heapSize = 0;
+
+    size_t   public_key_size = (grp->nbits+7)/8 ;
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
+    uint32_t secret_size =  ( ( grp->nbits + 7 ) / 8 ) ;
+    const uint32_t secret_size_in_heap = secret_size;
+    uint8_t* secret = mbedtls_calloc( 1, secret_size_in_heap );
```

Comment:
I dont think this will be a problem, but I would like to point out that this is implicitly casting from uint32_t to size_t.

Reply:
I don't follow you. Do you mean `secret_size_in_heap` ? This isn't a problem. `uint32_t` is either same size or smaller than `size_t`

Reply:
Yes. I dont think its a problem either, but note that `size_t` is not necessarily larger than `uint32_t`. This depends on the system and the C standard says what is the minimum max for size_t (which is smaller than 32 bits), but does not specify a max